easybuild/framework/easyconfig/types.py

```diff
@@ -602,32 +630,21 @@ def ensure_iterable_license_specs(specs):
     'opt_keys': ['full_mod_name', 'short_mod_name', 'toolchain', 'versionsuffix'],
     'req_keys': ['name', 'version'],
 }))
-DEPENDENCIES = (list, as_hashable({'elem_types': [DEPENDENCY_DICT]}))
+DEPENDENCY_TUPLE = (tuple, as_hashable({'elem_types': [str, TUPLE_OF_STRINGS, dict, bool]}))
+
+DEPENDENCY_TYPES = [DEPENDENCY_DICT, Dependency, DEPENDENCY_TUPLE]
+# For iterated dependencies
+DEPENDENCY_LIST = (list, as_hashable({'elem_types': DEPENDENCY_TYPES}))
+DEPENDENCIES = (list, as_hashable({'elem_types': DEPENDENCY_TYPES + [DEPENDENCY_LIST]}))
 
-TUPLE_OF_STRINGS = (tuple, as_hashable({'elem_types': [str]}))
-LIST_OF_STRINGS = (list, as_hashable({'elem_types': [str]}))
-STRING_OR_TUPLE_LIST = (list, as_hashable({'elem_types': [str, TUPLE_OF_STRINGS]}))
-STRING_DICT = (dict, as_hashable(
-    {
-        'elem_types': [str],
-        'key_types': [str],
-    }
-))
-STRING_OR_TUPLE_DICT = (dict, as_hashable(
-    {
-        'elem_types': [str],
-        'key_types': [str, TUPLE_OF_STRINGS],
-    }
-))
-STRING_OR_TUPLE_OR_DICT_LIST = (list, as_hashable({'elem_types': [str, TUPLE_OF_STRINGS, STRING_DICT]}))
 SANITY_CHECK_PATHS_ENTRY = (list, as_hashable({'elem_types': [str, TUPLE_OF_STRINGS, STRING_OR_TUPLE_DICT]}))
 SANITY_CHECK_PATHS_DICT = (dict, as_hashable({
     'elem_types': {
         SANITY_CHECK_PATHS_FILES: [SANITY_CHECK_PATHS_ENTRY],
         SANITY_CHECK_PATHS_DIRS: [SANITY_CHECK_PATHS_ENTRY],
     },
-    'opt_keys': [],
-    'req_keys': [SANITY_CHECK_PATHS_FILES, SANITY_CHECK_PATHS_DIRS],
+    'opt_keys': [SANITY_CHECK_PATHS_FILES, SANITY_CHECK_PATHS_DIRS],
```

Comment:
This actually needs to be discussed: We currently allow missing dirs/files in sanity-check-paths and only fail it in CI.

We should decide where we want to handle this:
- Make both optional when running easybuild but fail on CI for contributions
- Make both fully optional and default them to empty lists in the code using it (sanity_check_step)
- Make them required here (as before) and fill them with empty lists in the "conversion" function if not set, then we don't need to check if the keys exist in sanity_check_step. Optionally enforce non-empty lists on CI